marlowe/app/Benchmark/Marlowe.hs

```diff
@@ -0,0 +1,170 @@
+
+-----------------------------------------------------------------------------
+--
+-- Module      :  $Headers
+-- License     :  Apache 2.0
+--
+-- Stability   :  Experimental
+-- Portability :  Portable
+--
+-- | Benchmarking support for Marlowe's validators.
+--
+-----------------------------------------------------------------------------
+
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+
+
+module Benchmark.Marlowe
+  ( -- * Benchmarking
+    evaluationContext
+  , executeBenchmark
+  , printBenchmark
+  , printResult
+  , readBenchmark
+  , readBenchmarks
+  , tabulateResults
+  ) where
+
+
+import Benchmark.Marlowe.Types (Benchmark(..))
+import Codec.Serialise (deserialise)
+import Control.Monad.Except (runExcept)
+import Data.Bifunctor (first)
+import Data.Maybe (fromJust)
+import Language.Marlowe.Core.V1.Semantics (MarloweData)
+import Language.Marlowe.Scripts (MarloweInput)
+import Paths_marlowe_cardano (getDataDir)
+import Plutus.V2.Ledger.Api
+  ( Data(Constr, I)
+  , EvaluationContext
+  , EvaluationError
+  , ExBudget(ExBudget, exBudgetCPU, exBudgetMemory)
+  , ExCPU(ExCPU)
+  , ExMemory(ExMemory)
+  , LogOutput
+  , ProtocolVersion(..)
+  , ScriptContext(scriptContextTxInfo)
+  , SerializedScript
+  , TxInfo(txInfoId)
+  , ValidatorHash
+  , VerboseMode(Verbose)
+  , evaluateScriptCounting
+  , fromData
+  , mkEvaluationContext
+  , toData
+  )
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultCostModelParams)
+import System.Directory (listDirectory)
+import System.FilePath ((</>))
+
+import qualified Data.ByteString.Lazy as LBS (readFile)
+
+
+-- | Run a benchmark case.
+executeBenchmark
+  :: SerializedScript  -- ^ The serialised validator.
+  -> Benchmark  -- ^ The benchmarking case.
+  -> Either String (LogOutput, Either EvaluationError ExBudget)  -- ^ An error or the cost.
+executeBenchmark serialisedValidator Benchmark{..} =
+  case evaluationContext of
+   Left message -> Left message
+   Right ec ->
+    Right
+      $ evaluateScriptCounting (ProtocolVersion 8 0) Verbose ec serialisedValidator
+        [bDatum, bRedeemer, toData bScriptContext]
+
+
+-- | The execution context for benchmarking.
+evaluationContext :: Either String EvaluationContext
+evaluationContext =
+  let
+    costParams = fromJust defaultCostModelParams
+  in
+    first show . runExcept $ mkEvaluationContext costParams
+
+
+-- | Read all of the benchmarking cases for a particular validator.
+readBenchmarks
+  :: FilePath
+  -> IO (Either String [Benchmark])
+readBenchmarks subfolder =
+  do
+    folder <- (</> subfolder) <$> getDataDir :: IO FilePath
+    files <- fmap (folder </>) <$> listDirectory folder :: IO [FilePath]
+    sequence <$> mapM readBenchmark files
+
+
+-- | Read a benchmarking file.
+readBenchmark
+  :: FilePath
+  -> IO (Either String Benchmark)
+readBenchmark filename =
+  do
+    payload <- LBS.readFile filename
+    pure
+      $ case deserialise payload of
```

Comment:
I used `Data` because it is generic and can apply to any validator (e.g., semantics or role payout). These are portable enough that we're using them on a plutus branch, where they are deserialized, adjusted (because of the different script hashes there), and then used as benchmarks for changes in the plutus compiler.

I don't envision that the benchmark files would ever be updated unless the specification for the Marlowe validator fails.